src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -0,0 +1,529 @@
+ *
+ * and, access modes in the next 3 bits:
+ *  - 3rd bit is set for PARALLEL SELECT accesses to a relation
+ *  - 4th bit is set for PARALLEL DML accesses to a relation
+ *  - 5th bit is set for PARALLEL DDL accesses to a relation
+ *
+ */
+typedef struct RelationAccessHashKey
+{
+	Oid relationId;
+} RelationAccessHashKey;
+
+typedef struct RelationAccessHashEntry
+{
+	RelationAccessHashKey key;
+
+	int relationAccessMode;
+} RelationAccessHashEntry;
+
+static HTAB *RelationAccessHash;
+
+
+static void RecordPlacementAccessToCache(Oid relationId,
+										 ShardPlacementAccessType accessType);
+static RelationAccessMode GetRelationAccessMode(Oid relationId,
+												ShardPlacementAccessType accessType);
+static void RecordParallelRelationAccess(Oid relationId, ShardPlacementAccessType
+										 placementAccess);
+static void RecordParallelRelationAccessToCache(Oid relationId,
+												ShardPlacementAccessType placementAccess);
+
+
+/*
+ * Empty RelationAccessHash, without destroying the hash table itself.
+ */
+void
+ResetRelationAccessHash()
+{
+	hash_delete_all(RelationAccessHash);
+}
+
+
+/*
+ * Allocate RelationAccessHash.
+ */
+void
+AllocateRelationAccessHash()
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(RelationAccessHashKey);
+	info.entrysize = sizeof(RelationAccessHashEntry);
+	info.hash = tag_hash;
+	info.hcxt = ConnectionContext;
+	hashFlags = (HASH_ELEM | HASH_BLOBS | HASH_CONTEXT);
+
+	RelationAccessHash = hash_create("citus connection cache (relationid)",
+									 8, &info, hashFlags);
+}
+
+
+/*
+ * AssociatePlacementAccessWithRelation associates the placement access to the
+ * distributed relation that the placement belongs to.
+ */
+void
+AssociatePlacementAccessWithRelation(ShardPlacement *placement,
+									 ShardPlacementAccessType accessType)
+{
+	uint64 shardId = INVALID_SHARD_ID;
+	Oid relationId = InvalidOid;
+
+	if (!ShouldRecordRelationAccess())
+	{
+		return;
+	}
+
+	shardId = placement->shardId;
+	relationId = RelationIdForShard(shardId);
+
+	/*
+	 * If a relation is partitioned, record accesses to all of its partitions as well.
+	 * We prefer to use PartitionedTableNoLock() because at this point the necessary
+	 * locks on the relation has already been acquired.
+	 */
+	if (PartitionedTableNoLock(relationId))
+	{
+		List *partitionList = PartitionList(relationId);
+		ShardInterval *shardInterval = LoadShardInterval(shardId);
+		int shardIndex = shardInterval->shardIndex;
+		ListCell *partitionCell = NULL;
+
+		foreach(partitionCell, partitionList)
+		{
+			Oid partitionOid = lfirst_oid(partitionCell);
+			uint64 partitionShardId = INVALID_SHARD_ID;
+			List *partitionShardPlacementList = NIL;
+			ListCell *partitionCell = NULL;
+
+			/*
+			 * During create_distributed_table, the partitions may not
+			 * have been created yet.
+			 */
+			if (!IsDistributedTable(partitionOid))
+			{
+				continue;
```

Comment:
Might be worth adding a comment about why not registering is ok (we'll still register when we distribute the partitions).

Reply:
added